homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,396 @@
+        """Initialize the view."""
+        self._store = store
+
+    async def get(self, request):
+        """Handle a request to get all registered devices."""
+        all_devices = []
+
+        hass = request.app['hass']
+
+        for key, device in hass.data[DOMAIN].items():
+            all_devices.append(safe_device(device))
+
+        return self.json(all_devices)
+
+    @RequestDataValidator(REGISTER_DEVICE_SCHEMA)
+    async def post(self, request, data):
+        """Handle the POST request for device registration."""
+        hass = request.app['hass']
+
+        fingerprint = data[ATTR_FINGERPRINT]
+
+        if fingerprint in hass.data[DOMAIN]:
+            return Response(status=409)
+
+        resp = {}
+
+        webhook_id = generate_secret()
+
+        data[CONF_WEBHOOK_ID] = resp[CONF_WEBHOOK_ID] = webhook_id
+
+        if data[ATTR_SUPPORTS_ENCRYPTION] and supports_encryption():
+            secret = generate_secret(16)
+
+            data[CONF_SECRET] = resp[CONF_SECRET] = secret
+
+        hass.data[DOMAIN][fingerprint] = data
+
+        try:
+            await self._store.async_save(hass.data[DOMAIN])
+        except HomeAssistantError:
+            return self.json_message("Error saving device.",
+                                     HTTP_INTERNAL_SERVER_ERROR)
+
+        webhook.async_register(hass, DOMAIN, 'Mobile app', webhook_id,
+                               handle_webhook)
+
+        return self.json(resp, status_code=HTTP_CREATED)
+
+
+class SingleDeviceView(HomeAssistantView):
+    """A view that represents a single device."""
+
+    url = '/api/mobile_app/devices/{webhook_id}'
+    name = 'api:mobile_app:single-device'
+
+    def __init__(self, store):
+        """Initialize the view."""
+        self._store = store
+
+    async def get(self, request, webhook_id):
+        """Return what we know about the device."""
+        device = device_for_webhook_id(request.app['hass'], webhook_id)
+
+        if device is None:
+            return Response(status=404)
+
+        return self.json(safe_device(device))
+
+    @RequestDataValidator(REGISTER_DEVICE_SCHEMA)
+    async def put(self, request, webhook_id, data):
+        """Handle the update data request for device registration."""
+        hass = request.app['hass']
+
+        device = device_for_webhook_id(hass, webhook_id)
+
+        if device is None:
+            return Response(status=404)
+
+        fingerprint = device[ATTR_FINGERPRINT]
+
+        data[CONF_WEBHOOK_ID] = device[CONF_WEBHOOK_ID]
+        data[CONF_SECRET] = device[CONF_SECRET]
+
+        hass.data[DOMAIN][fingerprint] = data
+
+        try:
+            await self._store.async_save(hass.data[DOMAIN])
+        except HomeAssistantError:
+            return self.json_message("Error updating device.",
+                                     HTTP_INTERNAL_SERVER_ERROR)
+
+        return self.json(safe_device(data))
+
+
+config_entry_flow.register_discovery_flow(
```

Comment:
I don't think that this is necessary. Instead we add it to `default_config` or ask people to add it to their configuration if they don't use `default_config`.